src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Cohost/CohostProjectSnapshot.cs

```diff
@@ -0,0 +1,160 @@
+        _telemetryReporter = telemetryReporter;
+        _projectKey = ProjectKey.From(_project)!.Value;
+
+        _lazyConfiguration = new Lazy<RazorConfiguration>(CreateRazorConfiguration);
+        _lazyProjectEngine = new Lazy<RazorProjectEngine>(() => DefaultProjectEngineFactory.Create(
+            Configuration,
+            fileSystem: RazorProjectFileSystem.Create(Path.GetDirectoryName(FilePath)),
+            configure: builder =>
+            {
+                builder.SetRootNamespace(RootNamespace);
+                builder.SetCSharpLanguageVersion(CSharpLanguageVersion);
+                builder.SetSupportLocalizedComponentNames();
+            },
+            fallback: s_fallbackProjectEngineFactory,
+            factories: s_projectEngineFactories));
+
+        _tagHelpersLazy = new AsyncLazy<ImmutableArray<TagHelperDescriptor>>(() =>
+        {
+            var resolver = new CompilationTagHelperResolver(_telemetryReporter);
+            return resolver.GetTagHelpersAsync(_project, GetProjectEngine(), CancellationToken.None).AsTask();
+        }, joinableTaskContext.Factory);
+
+        _importsToRelatedDocumentsLazy = new Lazy<ImmutableDictionary<string, ImmutableArray<string>>>(() =>
+        {
+            var importsToRelatedDocuments = ImmutableDictionary.Create<string, ImmutableArray<string>>(FilePathNormalizer.Comparer);
+            foreach (var document in DocumentFilePaths)
+            {
+                var importTargetPaths = ProjectState.GetImportDocumentTargetPaths(document, FileKinds.GetFileKindFromFilePath(document), GetProjectEngine());
+                importsToRelatedDocuments = ProjectState.AddToImportsToRelatedDocuments(importsToRelatedDocuments, document, importTargetPaths);
+            }
+
+            return importsToRelatedDocuments;
+        });
+    }
+
+    public ProjectKey Key => _projectKey;
+
+    public RazorConfiguration Configuration => _lazyConfiguration.Value;
+
+    public IEnumerable<string> DocumentFilePaths
+        => _project.AdditionalDocuments
+            .Where(d => d.FilePath!.EndsWith(".razor", StringComparison.OrdinalIgnoreCase) || d.FilePath.EndsWith(".cshtml", StringComparison.OrdinalIgnoreCase))
+            .Select(d => d.FilePath.AssumeNotNull());
+
+    public string FilePath => _project.FilePath!;
+
+    public string IntermediateOutputPath => FilePathNormalizer.GetNormalizedDirectoryName(_project.CompilationOutputInfo.AssemblyPath);
+
+    public string? RootNamespace => _project.DefaultNamespace ?? "ASP";
+
+    public string DisplayName => _project.Name;
+
+    public VersionStamp Version => _project.Version;
+
+    public LanguageVersion CSharpLanguageVersion => ((CSharpParseOptions)_project.ParseOptions!).LanguageVersion;
+
+    public ImmutableArray<TagHelperDescriptor> TagHelpers => _tagHelpersLazy.GetValue();
+
+    public ProjectWorkspaceState ProjectWorkspaceState => ProjectWorkspaceState.Create(TagHelpers, CSharpLanguageVersion);
```

Comment:
Should this cache the `ProjectWorkspaceState` rather than constructing a new one each time the property is accessed? I don't think `TagHelpers` and `LanguageVersion` will change once their computed.